explainer.md

```diff
@@ -205,21 +209,213 @@ const config_with_namespaces = {
+use the special string `"*"`.
+
+```js
+const config_div_without_anything = {
+  elements: [ { name: "div", attributes: [] } ]
+};
+const config_div_with_everything = {
+  elements: [ { name: "div", attributes: "*" } ]
+};
+
+// I guess one could also write `removeAttributes: []`. Not sure if that's nicer.
+```
+
+### Comments
+
+Handling of HTML comment nodes can be controlled by an option. Setting
+`allowComments` to `true` allows them:
+
+```js
+const config_comments: { allow_comments: true };
+element.setHTML("XXX<!-- Hello world! -->XXX", {sanitizer: config_comments});
+// <div>XXX<!-- Hello world! -->XXX</div>
+```
+
+### Shadow Roots
+
+Declarative Shadow Roots are an HTML parser feature that parses `<template>`
+elements with `shadowrootmode` attributes and attaches the result as a
+shadow root to its parent elements. If this is not desired the
+`allowShadowRoots` attribute can be set to `false`. In either case, filtering
+rules will apply to the shadow roots as well.
+
+```js
+const example_with_shadow_root = "<template shadowrootmode=open><b onlick='alert(1)'>In the shadows.</b></template>";
+
+element.setHTML(example_with_shadow_root);
+//<div>
+//  #shadow-root: <b>In the shadows.</b>
+
+element.setHTMLUnsafe(example_with_shadow_root);
+//<div>
+//  #shadow-root: <b onlick='alert(1)'>In the shadows.</b>
+
+element.setHTML(example_with_shadow_root, {sanitizer: {allowShadowRoot: false}});
+// <div>
+
+element.setHTMLUnsafe(example_with_shadow_root, {sanitizer: {allowShadowRoot: false}});
+// <div>
+```
+
+### Configuration Errors
+
+The configuration allows expressing redundant or even contradictory options.
+For example, allowing and removing the same element. In cases where the
+meaning of a configuration dictionary isn't clear, we will
+throw a `TypeError` instead of making a best effort attempt at interpreting
+the configuration. A well-formed configuration has the following properties:
+
+
+* It contains either an allow-list or a remove-list, but not both.
+  * This applies to both element and attribute lists, seperately.
+  * Note that any config with both, an allow-list and remove-list, can be
+    rewritten by removing the remove-list items from the allow-list and then
+    droping the remove-list entirely.
+  * A config with an allow-list and a flatten-list makes sense since the items
+    in the flatten list preserve their child contents, while the allow-list does
+    not.
+* Any allow-, remove-, or flatten-list should contain each name at most once.
+  * This would apply to short forms as well.
+    E.g., `["div", { name: "div", namespace: "http://www.w3.org/1999/xhtml" }]`
+    contains the same name twice and would thus throw.
+  * While lists with duplicate element or attribute names could be coalesced,
+    it is ambiguous what the meaning of duplicate elements with different
+    element-dependent attribute lists would be.
+* The name must be set. If an `attributes` or `removeAttributes` key is present,
```

Comment:
Nit: s/name/element name/

Reply:
This was meant as applying to both element and attribute lists, so that a dictionary without a name-key set would be invalid.

Reply:
Fair enough. Keep it :)